infrastructure/src/main/java/org/corfudb/infrastructure/CompactorService.java

```diff
@@ -0,0 +1,163 @@
+        this.corfuStore = new CorfuStore(getCorfuRuntime());
+        this.compactionTriggerPolicy.setCorfuRuntime(getCorfuRuntime());
+        if (getCorfuRuntime().getParameters().getCheckpointTriggerFreqMillis() > 0) {
+            this.compactionTriggerFreqMs = getCorfuRuntime().getParameters().getCheckpointTriggerFreqMillis();
+        }
+
+        orchestratorThread.scheduleAtFixedRate(
+            this::runOrchestrator,
+            interval.toMillis(),
+            interval.toMillis(),
+            TimeUnit.MILLISECONDS
+        );
+    }
+
+    /**
+     * Invokes and cleans up the CorfuStoreCompactor jvm based on the status of CompactionManager
+     * Additionally, If the current node is the leader,
+     *      a. Invokes ValidateLiveness() to keep track of checkpointing progress by each client
+     *      b. Triggers the distributed compaction cycle based on the TriggerPolicy
+     */
+    private void runOrchestrator() {
+        boolean isLeader = isNodePrimarySequencer(updateLayoutAndGet());
+        compactorLeaderServices.setLeader(isLeader);
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            CheckpointingStatus managerStatus = (CheckpointingStatus) txn.getRecord(
+                    DistributedCompactor.COMPACTION_MANAGER_TABLE_NAME,
+                    DistributedCompactor.COMPACTION_MANAGER_KEY).getPayload();
+            txn.commit();
+            syslog.trace("ManagerStatus: {}", managerStatus == null ? "null" : managerStatus.getStatus());
+            if (managerStatus != null) {
+                if (managerStatus.getStatus() == StatusType.FAILED || managerStatus.getStatus() == StatusType.COMPLETED) {
+                    checkpointerJvmManager.shutdown();
+                    invokedJvm = false;
+                } else if (managerStatus.getStatus() == StatusType.STARTED_ALL && !checkpointerJvmManager.isRunning()
+                        && !invokedJvm) {
+                    checkpointerJvmManager.invokeCheckpointing();
+                    invokedJvm = true;
+                }
+            }
+
+            if (isLeader) {
+                if (managerStatus != null && (managerStatus.getStatus() == StatusType.STARTED ||
+                        managerStatus.getStatus() == StatusType.STARTED_ALL)) {
+                    compactorLeaderServices.validateLiveness(livenessTimeout);
+                } else if (compactionTriggerPolicy.shouldTrigger(this.compactionTriggerFreqMs)) {
+                    compactorLeaderServices.trimAndTriggerDistributedCheckpointing();
+                    compactionTriggerPolicy.markCompactionCycleStart();
+                }
+            }
+        } catch (Exception e) {
+            syslog.warn("Exception in runOrchestrator: {}", e.getStackTrace());
+        }
+    }
+
+    private Layout updateLayoutAndGet() {
+        Layout layout = getCorfuRuntime()
+                .invalidateLayout()
+                .join();
```

Comment:
would be great to have async code instead of blocking code

Reply:
even if this is made async. The caller of this method has to be blocking and cannot be async. So I think it might not be required to be made async.